gadgets/fsnotify/program.bpf.c

```diff
@@ -0,0 +1,523 @@
+// SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note
+/* Copyright (c) 2024 The Inspektor Gadget authors */
+
+#include <vmlinux.h>
+
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#include <gadget/buffer.h>
+#include <gadget/filesystem.h>
+#include <gadget/macros.h>
+#include <gadget/mntns_filter.h>
+#include <gadget/types.h>
+
+#define PATH_MAX 4096
+#define TASK_COMM_LEN 16
+
+enum type {
+	unknown,
+	dnotify,
+	inotify,
+	fanotify,
+	fa_resp,
+};
+
+enum fa_response {
+	na = 0,
+	allow = 0x01, // FAN_ALLOW
+	deny = 0x02, // FAN_DENY
+	interrupted = 0x03, // tracee interrupted (state != FAN_EVENT_ANSWERED)
+	// FAN_AUDIT and FAN_INFO not handled
+};
+
+struct enriched_event {
+	enum type type;
+
+	__u32 tracee_pid;
+	__u32 tracee_tid;
+	__u8 tracee_comm[TASK_COMM_LEN];
+	__u64 tracee_mntns_id;
+
+	__u32 tracer_pid;
+	__u32 tracer_tid;
+	__u8 tracer_comm[TASK_COMM_LEN];
+	gadget_mntns_id mntns_id;
+
+	__u32 prio;
+
+	__u32 fa_type;
+	__u32 fa_mask;
+	__u32 fa_pid;
+	__u32 fa_flags;
+	__u32 fa_f_flags;
+
+	__s32 i_wd;
+	__u32 i_mask;
+	__u32 i_cookie;
+	__u32 i_ino;
+	__u32 i_ino_dir;
+
+	__u8 name[PATH_MAX];
+};
+static const struct enriched_event empty_enriched_event = {};
+
+// context for the caller of fsnotify_insert_event
+struct fsnotify_insert_event_value {
+	enum type type;
+	__u32 i_ino;
+	__u32 i_ino_dir;
+};
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 64);
+	__type(key, u64); // tgid_pid
+	__type(value, struct fsnotify_insert_event_value);
+} fsnotify_insert_event_ctx SEC(".maps");
+
+// context for kprobe/kretprobe fsnotify_remove_first_event
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 64);
+	__type(key, u64); // tgid_pid
+	__type(value, void *); // struct fsnotify_group *
+} fsnotify_remove_first_event_ctx SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, void *); // struct fsnotify_event *event
+	__type(value, struct enriched_event);
+	__uint(max_entries, 10240);
+} enriched_fsnotify_events SEC(".maps");
+
+const volatile u64 tracer_group = 0;
+const volatile pid_t tracer_pid = 0;
+const volatile pid_t tracee_pid = 0;
+const volatile bool inotify_only = false;
+const volatile bool fanotify_only = false;
+GADGET_PARAM(tracer_group);
+GADGET_PARAM(tracer_pid);
+GADGET_PARAM(tracee_pid);
+GADGET_PARAM(inotify_only);
+GADGET_PARAM(fanotify_only);
+
+struct gadget_event {
```

Comment:
Please either add `_raw` suffix or use the `ebpf.formatter.enum` annotation for enumerations to work: https://www.inspektor-gadget.io/docs/latest/gadget-devel/gadget-ebpf-api#enumerations

```bash
WARN[0000] Failed to get target name for enum field "type": neither "ebpf.formatter.enum" annotation nor '_raw' suffix found
WARN[0000] Failed to get target name for enum field "fa_type": neither "ebpf.formatter.enum" annotation nor '_raw' suffix found
WARN[0000] Failed to get target name for enum field "fa_response": neither "ebpf.formatter.enum" annotation nor '_raw' suffix found
```